### PR TITLE
Handle ConnectionError on HTTP HEAD Requests

### LINKED ISSUE #9306: Linkcheck Reports Broken Link when Remote Server Closes on HEAD Request
**Describe the bug**
Running `make linkcheck` on a document that contains an external link to a website may report the link is broken when a web browser may successfully open the link. Specifically, if the website closes its connection when receiving the `HTTP HEAD` request method, then `linkcheck.py` will receive a `ConnectionError` exception, which bypasses the logic that would otherwise have it make an `HTTP GET` request. 

A specific example of a website exhibiting this behaviour is [the US Patent and Trademark Office](https://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO2&Sect2=HITOFF&p=1&u=%2Fnetahtml%2FPTO%2Fsearch-bool.html&r=1&f=G&l=50&co1=AND&d=PTXT&s1=7840660&OS=7840660&RS=7840660)

**To Reproduce**
Steps to reproduce the behaviour:
```sh
$ sphinx-quickstart  # accept all the default options
$ echo '\n\nThis is `a link to the US Patent Website <https://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO2&Sect2=HITOFF&p=1&u=%2Fnetahtml%2FPTO%2Fsearch-bool.html&r=1&f=G&l=50&co1=AND&d=PTXT&s1=7840660&OS=7840660&RS=7840660>`_.\n' >> index.rst
$ make html linkcheck 
```
- Observe linkcheck reporting a broken link:
```
Running Sphinx v4.0.2
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [linkcheck]: targets for 1 source files that are out of date
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
preparing documents... done
writing output... [100%] index

(           index: line   22) broken    https://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO2&Sect2=HITOFF&p=1&u=%2Fnetahtml%2FPTO%2Fsearch-bool.html&r=1&f=G&l=50&co1=AND&d=PTXT&s1=7840660&OS=7840660&RS=7840660 - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
build finished with problems.
make: *** [linkcheck] Error 1
```

- Open `_build/html/index.html` in your browser
- Click "a link to the US Patent Website"
- Observe the link opening and rendering normally

**Expected behavior**
If a link is valid, and the website is returning valid content for an `HTTP GET` request, `make linkcheck` should not report the link as `broken`.

Internally, in `sphinx/builders/linkcheck.py`, if a call to `requests.head()` raises a `requests.exceptions.ConnectionError` exception, it should attempt a `requests.get()` just like it does with `HTTPError` and `TooManyredirects`.

**Your project**
[sphinx-bug-linkcheck-reports-broken-link.zip](https://github.com/sphinx-doc/sphinx/files/6610252/sphinx-bug-linkcheck-reports-broken-link.zip)

**Environment info**
- OS: `macOS 11.3.1` (but this does not appear to be OS-dependent)
- Python version: `3.8.10` and `3.9.5`
- Sphinx version: `v3.5.4` and `v4.0.2`
- Sphinx extensions:  none
- Extra tools: any common web browser

**Additional context**
Using `curl`, we can see that this particular website closes connections when receiving `HTTP HEAD`:
```sh
$ curl -v -L -A "Sphinx/1.0" "http://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO1&Sect2=HITOFF&d=PALL&p=1&u=%2Fnetahtml%2FPTO%2Fsrchnum.htm&r=1&f=G&l=50&s1=9,942,040.PN.&OS=PN/9,942,040&RS=PN/9,942,040"
```
- Observe that this returns the expected HTML content

```sh
curl --head -v -L -A "Sphinx/1.0" "http://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO1&Sect2=HITOFF&d=PALL&p=1&u=%2Fnetahtml%2FPTO%2Fsrchnum.htm&r=1&f=G&l=50&s1=9,942,040.PN.&OS=PN/9,942,040&RS=PN/9,942,040"
```
- Observe that this fails to return any content from the server

### PR DESCRIPTION
**Pull request by jamathews**
Subject: Resolves #9306 

### Feature or Bugfix
- Bugfix

### Purpose
- #9306 is causing `make linkcheck` to fail on URLs that will open and render properly in a web browser. Prior to this fix, the link checker would incorrectly call these links broken.

### Detail
- The US Patent and Trademark website seems to close connections on `HTTP HEAD` requests, but will return the requested document for `HTTP GET` requests. The `linkcheck.py` code wasn't catching the `ConnectionError` exception, and as a result the code would skip the `requests.get()` attempt. With this fix, if `requests.head()` raises a `ConnectionError`, it will attempt a `requests.get()` which, for the UT Patent website, succeeds. If the website truly is rejecting all connections, an exception on the `requests.get()` call will still send control to the `except Exception as err` block as before.

### Relates
- #9306

### CODE REVIEW

File: sphinx/builders/linkcheck.py [hunk 1]
```diff
@@ -460,7 +460,7 @@ def check_uri() -> Tuple[str, str, int]:
                                                  config=self.config, auth=auth_info,
                                                  **kwargs)
                         response.raise_for_status()
-                    except (HTTPError, TooManyRedirects) as err:
+                    except (HTTPError, TooManyRedirects, ConnectionError) as err:
```
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]: IMO, a comment would help explaining `ConnectionError`. If the connection couldn’t be made to the server, the question “why retry as GET” is legitimate and a comment would help answering it.

File: tests/roots/test-linkcheck-connection-error-on-http-head/index.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+This tests bug #9306.
+
+This is `a link to the US Patent Website <https://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO2&Sect2=HITOFF&p=1&u=%2Fnetahtml%2FPTO%2Fsearch-bool.html&r=1&f=G&l=50&co1=AND&d=PTXT&s1=7840660&OS=7840660&RS=7840660>`_
```
Comment (francoisfreitag) [inline tests/roots/test-linkcheck-connection-error-on-http-head/index.rst]: I wonder if the behavior can be emulated with the `http_server` context manager?
My concern is that if this site changes behavior, the test will stop being useful.
That would be much faster (runs over localhost).

**Review by francoisfreitag — COMMENTED** [review]
Thanks for the nice report and patch! :+1:
I agree Sphinx could be more tolerant here.

---

Please add a note to the [CHANGES](https://github.com/sphinx-doc/sphinx/blob/4.x/CHANGES).

File: tests/roots/test-linkcheck-connection-error-on-http-head/index.rst (hunk 2 above)
Comment (jamathews) [reply tests/roots/test-linkcheck-connection-error-on-http-head/index.rst]: That's a good idea. I will attempt to replace the test with one that doesn't depend on an external service.

File: CHANGES [hunk 3]
```diff
@@ -16,6 +16,8 @@ Features added
 Bugs fixed
 ----------
 
+* #9306: Linkcheck Reports Broken Link when Remote Server Closes on HEAD Request
```
Comment (francoisfreitag) [inline CHANGES]: No need for title casing.
```suggestion
* #9306: Linkcheck reports broken link when remote server closes on HEAD request
```

File: sphinx/builders/linkcheck.py [hunk 4]
```diff
@@ -26,7 +26,7 @@
 from docutils import nodes
 from docutils.nodes import Element
 from requests import Response
-from requests.exceptions import HTTPError, TooManyRedirects
+from requests.exceptions import HTTPError, TooManyRedirects, ConnectionError
```
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]: Can keep alphabetical ordering.
```suggestion
from requests.exceptions import ConnectionError, HTTPError, TooManyRedirects
```

File: sphinx/builders/linkcheck.py [hunk 5]
```diff
@@ -460,7 +460,12 @@ def check_uri() -> Tuple[str, str, int]:
                                                  config=self.config, auth=auth_info,
                                                  **kwargs)
                         response.raise_for_status()
-                    except (HTTPError, TooManyRedirects) as err:
+                    # When there is a ConnectionError from the HEAD request, it might be due to
+                    # the webserver intentionally dropping the connection when it sees HEAD requests
+                    # but still responding properly to GET requests. This has been observed in the
+                    # wild with at https://patft.uspto.gov/.
+                    # See https://github.com/sphinx-doc/sphinx/issues/9306 for more details.
+                    except (HTTPError, TooManyRedirects, ConnectionError) as err:
```
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]:
```suggestion
                    except (ConnectionError, HTTPError, TooManyRedirects) as err:
```

File: sphinx/builders/linkcheck.py [hunk 6]
```diff
@@ -460,7 +460,12 @@ def check_uri() -> Tuple[str, str, int]:
                                                  config=self.config, auth=auth_info,
                                                  **kwargs)
                         response.raise_for_status()
-                    except (HTTPError, TooManyRedirects) as err:
+                    # When there is a ConnectionError from the HEAD request, it might be due to
+                    # the webserver intentionally dropping the connection when it sees HEAD requests
+                    # but still responding properly to GET requests. This has been observed in the
+                    # wild with at https://patft.uspto.gov/.
+                    # See https://github.com/sphinx-doc/sphinx/issues/9306 for more details.
```
Comment (francoisfreitag) [inline sphinx/builders/linkcheck.py]: The details are available through git blame, just a short comment is enough.
```suggestion
                    # Servers drop the connection on HEAD requests, causing
                    # ConnectionError.
```

File: tests/roots/test-linkcheck-connection-error-on-http-head/index.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+`This <https://localhost:7777>`_ tests bug #9306.
```
Comment (francoisfreitag) [inline tests/roots/test-linkcheck-connection-error-on-http-head/index.rst]: No need for an extra test case, can use the `test-linkcheck-localserver` document root.

File: tests/test_build_linkcheck.py [hunk 8]
```diff
@@ -575,3 +575,21 @@ def test_limit_rate_bails_out_after_waiting_max_time(app):
...
+
+    def do_GET(self):
+        self.send_response(200, "OK")
+        self.end_headers()
+
+
+@pytest.mark.sphinx('linkcheck', testroot='linkcheck-connection-error-on-http-head', freshenv=True)
+def test_get_after_head_raises_connection_error(app):
+    app.config.tls_verify = False
+    with https_server(ConnectionResetHandler):
```
Comment (francoisfreitag) [inline tests/test_build_linkcheck.py]:
```suggestion
    with http_server(ConnectionResetHandler):
```

File: tests/test_build_linkcheck.py [hunk 9]
```diff
@@ -575,3 +575,21 @@ def test_limit_rate_bails_out_after_waiting_max_time(app):
...
+class ConnectionResetHandler(http.server.BaseHTTPRequestHandler):
+    def do_HEAD(self):
+        raise requests.ConnectionError
+
+    def do_GET(self):
+        self.send_response(200, "OK")
+        self.end_headers()
+
+
+@pytest.mark.sphinx('linkcheck', testroot='linkcheck-connection-error-on-http-head', freshenv=True)
```
Comment (francoisfreitag) [inline tests/test_build_linkcheck.py]:
```suggestion
@pytest.mark.sphinx('linkcheck', testroot='linkcheck-localserver', freshenv=True)
```

File: tests/test_build_linkcheck.py [hunk 10]
```diff
@@ -575,3 +575,21 @@ def test_limit_rate_bails_out_after_waiting_max_time(app):
                                               rate_limits)
     next_check = worker.limit_rate(FakeResponse())
     assert next_check is None
+
+
+class ConnectionResetHandler(http.server.BaseHTTPRequestHandler):
+    def do_HEAD(self):
+        raise requests.ConnectionError
```
Comment (francoisfreitag) [inline tests/test_build_linkcheck.py]: The web server runs in a separate thread. I believe the reason the test works is that the unhandled exception causes the connection to be closed, but the raised exception does not propagate to the linkcheck builder thread.

Instead, be explicit about what the HTTP server is doing: closing the connection.
```suggestion
        self.connection.close()
```

File: tests/test_build_linkcheck.py [hunk 11]
```diff
@@ -575,3 +575,21 @@ def test_limit_rate_bails_out_after_waiting_max_time(app):
...
+        self.end_headers()
+
+
+@pytest.mark.sphinx('linkcheck', testroot='linkcheck-connection-error-on-http-head', freshenv=True)
+def test_get_after_head_raises_connection_error(app):
+    app.config.tls_verify = False
+    with https_server(ConnectionResetHandler):
+        app.build()
+    content = (app.outdir / 'output.txt').read_text()
+    assert not content
```
Comment (francoisfreitag) [inline tests/test_build_linkcheck.py:598]: Can also assert the content of the JSON to make sure link was considered as working.

```suggestion
    assert not content
    content = (app.outdir / 'output.json').read_text()
    assert json.loads(content) == {
        "filename": "index.rst",
        "lineno": 1,
        "status": "working",
        "code": 0,
        "uri": "http://localhost:7777/",
        "info": "",
    }
```

**Review by francoisfreitag — COMMENTED** [review]
Nice! A few minor changes and this should be good to go.

File: CHANGES [hunk 12]
```diff
@@ -16,6 +16,8 @@ Features added
 Bugs fixed
 ----------
 
+* #9306: Linkcheck reports broken link when remote server closes on HEAD request
```
Comment (francoisfreitag) [inline CHANGES]:
```suggestion
* #9306: Linkcheck reports broken link when remote server closes the connection
  on HEAD request
```

**Review by francoisfreitag — APPROVED** [review]
Looking good to me, thanks!

**Comment by tk0miya** [comment]
Thank you for your contribution. +1 for merging this into 4.x branch instead of 4.0.x. Usually, we use 4.0.x branch only for severe or critical bug fixes. So this is better to merge this into 4.x branch.

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
Merged. Thank you for your contribution!

@francoisfreitag Thank you for reviewing :-)

**Pull request merged**